### PR TITLE
Adding reactive programming introduction, issue #14

### PR DESCRIPTION
**Pull request by rorticus**
Introduction to reactive programming. Just basically saying how we use it in Dojo2. I asked around and it sounds like we don't really use Observables right now so I let them out.

Resolves: #14

### CODE REVIEW

File: site/source/tutorials/reactive_programming.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+Reactive programming is concenered with handling change, specifically, the _propagation of change_. As web applications have become more and more asynchronous, keeping track of application state has become increasingly complicated. With traditional programming patterns we rely on event listeners or try to keep track of application state manually.
+
+Consider a banking application where the user's account balance might show up in several different places on the screen. After the user makes a withdrawl, we want the account balance to update.
+
+## Imperative Approach
+
+Traditionally, we might use an event emitter. Every time balance is updated we emit an event, and any interested parties can listen in.
+
+![Banking](resources/imperative.png)
```
Comment (rorticus) [inline site/source/tutorials/reactive_programming.md]: Not sure where we put images...

File: site/source/tutorials/reactive_programming.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
+# Reactive Programming
```
Comment (dylans) [inline site/source/tutorials/reactive_programming.md]: Reactive programming (we use lowercase for titles on the Dojo 2 site for non-first words unless they are proper nouns

File: site/source/tutorials/reactive_programming.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+Reactive programming is concenered with handling change, specifically, the _propagation of change_. As web applications have become more and more asynchronous, keeping track of application state has become increasingly complicated. With traditional programming patterns we rely on event listeners or try to keep track of application state manually.
```
Comment (dylans) [inline site/source/tutorials/reactive_programming.md]: increasingly complicated -> increasingly complex

Comment (dylans) [inline site/source/tutorials/reactive_programming.md]: patterns we rely on event listeners or try to keep track of application state manually -> patterns, we traditionally rely on event listeners or try manually track application state

File: site/source/tutorials/reactive_programming.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+Consider a banking application where the user's account balance might show up in several different places on the screen. After the user makes a withdrawl, we want the account balance to update.
```
Comment (dylans) [inline site/source/tutorials/reactive_programming.md]: show up -> appear

File: site/source/tutorials/reactive_programming.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+## Imperative Approach
+
+Traditionally, we might use an event emitter. Every time balance is updated we emit an event, and any interested parties can listen in.
```
Comment (dylans) [inline site/source/tutorials/reactive_programming.md]: Every time balance is updated we emit an event, and any interested parties can listen in. -> Every time the balance is updated, we emit an event, and any interested parties can listen for these events.

File: site/source/tutorials/reactive_programming.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+});
+```
+
+With an event emitter, any interested widget can listen for updates to the account balance.  This pattern has served well in the past, but there are a few problems with this approach:
+
+*   We need to remember to emit `account-balance-updated` every time the account balance is changed.
+*   We need to rememer to listen for balance updates.
+*   We need to have a facitlity to **remove** the event listeners when they are no longer needed.
+
+## Reactive Approach
```
Comment (dylans) [inline site/source/tutorials/reactive_programming.md]: Reactive approach

File: site/source/tutorials/reactive_programming.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+Dojo2 is built around unidirectional, top-down ,property propagation It is the parent widget's job to pass properties down to it's children. In fact, a child widget has no way to reference a parent widget! When a property changes, widgets are re-rendered (using an efficient virtual DOM) to reflect the updated state.
```
Comment (dylans) [inline site/source/tutorials/reactive_programming.md]: Dojo2 -> Dojo 2

Comment (dylans) [inline site/source/tutorials/reactive_programming.md]: unidirectional, top-down ,property propagation  -> uni-directional, top-down property propagation.

Comment (dylans) [inline site/source/tutorials/reactive_programming.md]: it's children -> its children

Comment (dylans) [inline site/source/tutorials/reactive_programming.md]: no way to reference a -> no direct reference to a ?

File: site/source/tutorials/reactive_programming.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+*   We need to remember to emit `account-balance-updated` every time the account balance is changed.
+*   We need to rememer to listen for balance updates.
+*   We need to have a facitlity to **remove** the event listeners when they are no longer needed.
+
+## Reactive Approach
+
+Dojo2 is built around unidirectional, top-down ,property propagation It is the parent widget's job to pass properties down to it's children. In fact, a child widget has no way to reference a parent widget! When a property changes, widgets are re-rendered (using an efficient virtual DOM) to reflect the updated state. 
+
+In our banking example, we simply provide the account balance as a property to the widget, and when the property changes, so does our UI. The account balance is propagated down to the widgets that need it.
```
Comment (dylans) [inline site/source/tutorials/reactive_programming.md]: UI -> user interface

File: site/source/tutorials/reactive_programming.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+class AccountBalance extends WidgetBase<AccountBalanceProperties> {
+    render() {
+      	const { balance } = this.properties;
+        return v('span', {}, [ balance ]);
+    }
+}
+```
+
+Our `AccountBalance` widget is only concerned with it's `balance` property, and that property is provided by the parent widget.  Setting the balance property at the application level will cause the balance to be updated across the entire application.
```
Comment (dylans) [inline site/source/tutorials/reactive_programming.md]: with it's -> with its

File: site/source/tutorials/reactive_programming.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+## What's Next?
```
Comment (dylans) [inline site/source/tutorials/reactive_programming.md]: What's next?

File: site/source/tutorials/reactive_programming.md [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+Reactive programming patterns are being used in Dojo2 to make widgets consistent and easy to author.  Want to learn more about reactive programming? There are a number of additional reactive patterns that can be used to manage your application state (like [Redux](http://redux.js.org/), or [Observables](http://reactivex.io/documentation/observable.html)).
```
Comment (dylans) [inline site/source/tutorials/reactive_programming.md]: Dojo2 -> Dojo 2

Comment (dylans) [inline site/source/tutorials/reactive_programming.md]: May want to add a sentence that says that Dojo 2 will also soon have a reactive application store that is expected to arrive for beta 3?

**Comment by rorticus** [comment]
@dylans Updated!

**Comment by dylans** [comment]
Approved, though I believe this needs layout metadata added, which @vansimke can help with.

**Comment by vansimke** [comment]
@rorticus you should be able to use the first 6 lines of the [TypeScript and Dojo2](https://raw.githubusercontent.com/dojo/dojo.io/master/site/source/tutorials/TypeScript_and_Dojo_2.md) article as a template for the front-matter for this article.

**Comment by rorticus** [comment]
Thanks @vansimke ! I stole that link and wrote a quick `overview`

**Comment by rorticus** [comment]
@pottedmeat I've made your suggested changes! 👍

File: site/source/tutorials/reactive_programming.md [hunk 12]
```diff
@@ -1,25 +1,25 @@
...
-overview: Dojo 2 embracees reactive programming patterns like top-down property propagation. This article gives a high level overview of the reactive patterns in Dojo 2 and how they are different from more traditional patterns.
+overview: Dojo 2 embraces reactive programming patterns like top-down property propagation. This article gives a high level overview of the reactive patterns in Dojo 2 and how they are different from more traditional patterns.
 ---
 # Reactive programming
 
-Reactive programming is concenered with handling change, specifically, the _propagation of change_. As web applications have become more and more asynchronous, keeping track of application state has become increasingly complex. With traditional programming patterns, we rely on event listeners or try to keep track of application state manually.
+Reactive programming is concernedwith handling change - specifically, the _propagation of change_. As web applications have become more and more asynchronous, keeping track of application state has become increasingly complex. With traditional programming patterns, we rely on event listeners or try to keep track of application state manually.
```
Comment (agubler) [inline site/source/tutorials/reactive_programming.md]: concernedwith -> concerned with

File: site/source/tutorials/reactive_programming.md [hunk 13]
```diff
@@ -0,0 +1,63 @@
+---
+layout: false
+title: Reactive programming
+overview: Dojo 2 embraces reactive programming patterns like top-down property propagation. This article gives a high level overview of the reactive patterns in Dojo 2 and how they are different from more traditional patterns.
```
Comment (agubler) [inline site/source/tutorials/reactive_programming.md]: maybe unidirectional over top-down?

File: site/source/tutorials/reactive_programming.md [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+---
+# Reactive programming
+
+Reactive programming is concernedwith handling change - specifically, the _propagation of change_. As web applications have become more and more asynchronous, keeping track of application state has become increasingly complex. With traditional programming patterns, we rely on event listeners or try to keep track of application state manually.
```
Comment (agubler) [inline site/source/tutorials/reactive_programming.md]: not sure about "more and more asynchronous", can we be more specific about the challenges (i don't think it's just asynchronous flow)

File: site/source/tutorials/reactive_programming.md [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+Consider a banking application where the user's account balance might appear in several different places on the screen. After the user makes a withdrawal, we want the account balance to update.
+
+## Imperative approach
+
+Traditionally, we might use an event emitter. Each time the balance is updated we emit an event, and any interested parties can listen for these events.
+
+![Banking](resources/imperative.png)
+
+```javascript
```
Comment (agubler) [inline site/source/tutorials/reactive_programming.md]: not sure it makes a difference but usually we do typescript

File: site/source/tutorials/reactive_programming.md [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+With an event emitter, any interested widget can listen for updates to the account balance.  This pattern has served well in the past, but there are a few problems with this approach:
+
+*   We need to remember to emit `account-balance-updated` every time the account balance is changed.
+*   We need to remember to listen for balance updates.
+*   We need to have a facility to **remove** the event listeners when they are no longer needed.
+
+## Reactive approach
+
+Dojo 2 is built around unidirectional, top-down property propagation It is the parent widget's job to pass properties down to its children. In fact, a child widget has no direct reference to a parent widget! When a property changes, widgets are re-rendered (using an efficient virtual DOM) to reflect the updated state.
```
Comment (agubler) [inline site/source/tutorials/reactive_programming.md]: It -> it

File: site/source/tutorials/reactive_programming.md [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+## Reactive approach
+
+Dojo 2 is built around unidirectional, top-down property propagation It is the parent widget's job to pass properties down to its children. In fact, a child widget has no direct reference to a parent widget! When a property changes, widgets are re-rendered (using an efficient virtual DOM) to reflect the updated state. 
+
+In our banking example, we simply provide the account balance as a property to the widget - and when the property changes, so does our user interface. The account balance is propagated down to the widgets that need it.
+
+![Account Balance](resources/dojo2-reactive.png)
+
+```typescript
+export interface AccountBalanceProperties extends WidgetProperties {
```
Comment (agubler) [inline site/source/tutorials/reactive_programming.md]: don't need to extend `WidgetProperties`

File: site/source/tutorials/reactive_programming.md [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+    balance: number;
+}
+
+class AccountBalance extends WidgetBase<AccountBalanceProperties> {
+    render() {
```
Comment (agubler) [inline site/source/tutorials/reactive_programming.md]: should show this is `protected` as not doing so this makes it `public`

File: site/source/tutorials/reactive_programming.md (hunk 16 above)
Comment (dylans) [reply site/source/tutorials/reactive_programming.md]: propagation It is -> propagation where It is

**Comment by rorticus** [comment]
@agubler your feedback has been addressed except for the "more and more asynchronous" bit. I agree we can be a little bit more specific. Any thoughts on what challenges you were thinking of?

**Comment by agubler** [comment]
@rorticus I don't, if no one else has any other ideas then I guess we can leave it in!

Ping to @pottedmeat for a re-review :)

File: site/source/tutorials/reactive_programming.md [hunk 19]
```diff
@@ -0,0 +1,63 @@
+---
+layout: false
+title: Reactive programming
+overview: Dojo 2 embraces reactive programming patterns like unidirectional property propagation. This article gives a high level overview of the reactive patterns in Dojo 2 and how they are different from more traditional patterns.
+---
+# Reactive programming
+
+Reactive programming is concernedwith handling change - specifically, the _propagation of change_. As web applications have become more and more asynchronous, keeping track of application state has become increasingly complex. With traditional programming patterns, we rely on event listeners or try to keep track of application state manually.
```
Comment (vansimke) [inline site/source/tutorials/reactive_programming.md]: missing space: concernedwith -> concerned with

**Comment by agubler** [comment]
@rorticus could you quickly fix up the last remaining comment from @vansimke? Thanks!

**Comment by rorticus** [comment]
@agubler @vansimke typo fixed!

**Pull request merged**